File: src/types/domain.ts
export type EntityId = string;
export type IsoDateTime = string;

export const TEST_CASE_TYPES = ["api", "ui"] as const;
export type TestCaseType = (typeof TEST_CASE_TYPES)[number];

export const RUN_STATUSES = [
  "idle",
  "queued",
  "running",
  "passed",
  "failed",
  "cancelled"
] as const;
export type RunStatus = (typeof RUN_STATUSES)[number];

export const BROWSER_RUNTIME_STATUSES = ["healthy", "degraded", "unavailable"] as const;
export type BrowserRuntimeStatus = (typeof BROWSER_RUNTIME_STATUSES)[number];

export const RECORDING_STATUSES = ["idle", "recording", "stopped", "failed"] as const;
export type RecordingStatus = (typeof RECORDING_STATUSES)[number];

export const REPLAY_STATUSES = ["idle", "running", "passed", "failed", "cancelled"] as const;
export type ReplayStatus = (typeof REPLAY_STATUSES)[number];

export const VARIABLE_KINDS = ["plain", "secret"] as const;
export type VariableKind = (typeof VARIABLE_KINDS)[number];

export const ASSERTION_OPERATORS = [
  "status_equals",
  "json_path_exists",
  "json_path_equals",
  "body_contains",
  "header_equals"
] as const;
export type AssertionOperator = (typeof ASSERTION_OPERATORS)[number];

export const STEP_ACTIONS = [
  "navigate",
  "click",
  "fill",
  "select",
  "check",
  "uncheck",
  "wait_for",
  "assert_text"
] as const;
export type StepAction = (typeof STEP_ACTIONS)[number];

File: src-tauri/src/contracts/domain.rs
use serde::{Deserialize, Serialize};

pub type EntityId = String;
pub type IsoDateTime = String;

#[derive(Debug, Clone, Copy, Serialize, Deserialize, PartialEq, Eq)]
#[serde(rename_all = "snake_case")]
pub enum EnvironmentType {
    Development,
    Staging,
    Production,
    Custom,
}

#[derive(Debug, Clone, Copy, Serialize, Deserialize, PartialEq, Eq)]
#[serde(rename_all = "snake_case")]
pub enum TestCaseType {
    Api,
    Ui,
}

#[derive(Debug, Clone, Copy, Serialize, Deserialize, PartialEq, Eq)]
#[serde(rename_all = "snake_case")]
pub enum RunStatus {
    Idle,
    Queued,
    Running,
    Passed,
    Failed,
    Cancelled,
}

#[derive(Debug, Clone, Copy, Serialize, Deserialize, PartialEq, Eq)]
#[serde(rename_all = "snake_case")]
pub enum BrowserRuntimeStatus {
    Healthy,
    Degraded,
    Unavailable,
}

#[derive(Debug, Clone, Copy, Serialize, Deserialize, PartialEq, Eq)]
#[serde(rename_all = "snake_case")]
pub enum RecordingStatus {
    Idle,
    Recording,
    Stopped,
    Failed,
}

#[derive(Debug, Clone, Copy, Serialize, Deserialize, PartialEq, Eq)]
#[serde(rename_all = "snake_case")]
pub enum ReplayStatus {
    Idle,
    Running,
    Passed,
    Failed,
    Cancelled,
}

#[derive(Debug, Clone, Copy, Serialize, Deserialize, PartialEq, Eq)]
#[serde(rename_all = "snake_case")]
pub enum VariableKind {
    Plain,
    Secret,
}

#[derive(Debug, Clone, Copy, Serialize, Deserialize, PartialEq, Eq)]
#[serde(rename_all = "snake_case")]
pub enum AssertionOperator {
    StatusEquals,
    JsonPathExists,
    JsonPathEquals,
    BodyContains,
    HeaderEquals,
}

#[derive(Debug, Clone, Copy, Serialize, Deserialize, PartialEq, Eq)]
#[serde(rename_all = "snake_case")]
pub enum StepAction {
    Navigate,
    Click,
    Fill,
    Select,
    Check,
    Uncheck,
    WaitFor,
    AssertText,
}

#[derive(Debug, Clone, Copy, Serialize, Deserialize, PartialEq, Eq)]
#[serde(rename_all = "snake_case")]
pub enum StepConfidence {
    High,
    Medium,
    Low,
}

#[cfg(test)]
mod tests {
    use super::{AssertionOperator, StepAction};

    #[test]
    fn serializes_assertion_operator_with_canonical_contract_name() {
        let value = serde_json::to_value(AssertionOperator::StatusEquals)
            .expect("failed to serialize assertion operator");

        assert_eq!(
            value,
            serde_json::Value::String("status_equals".to_string())
        );
    }

    #[test]
    fn serializes_step_action_with_canonical_contract_name() {
        let value =
            serde_json::to_value(StepAction::WaitFor).expect("failed to serialize step action");

        assert_eq!(value, serde_json::Value::String("wait_for".to_string()));
    }
}
